### PR TITLE
allow unused input parameters passthrough when chunking in asr pipelines

### PR DESCRIPTION
**Pull request by VictorAtIfInsurance**
# What does this PR do?

Harmonizing behaviour for what input data is passthrough automatic speech recognition pipelines.

Today you can get different return dictionaries keys when using ASR pipeline with and without the call parameter `chunk_length_s`. With this parameter you will only get a dictionary with the transformed input data, while omitting this parameter allows you to retrieve any unused input data parameter back in the resulting dictionary. This inconsistency can be confusing as sometimes you need to keep track of which inputs corresponds to which outputs, while it is not obvious to a user why.

The following code snippet highlights these differences:

```
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
import numpy as np


device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v3-turbo"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)

data = {"raw": np.zeros(100), "sampling_rate": 16_000, "id": 1}

chunk_output = pipe(data.copy(), chunk_length_s=30)
non_chunk_output = pipe(data.copy())

print(chunk_output)
print(non_chunk_output)
assert chunk_output.keys() == non_chunk_output.keys()
```

This will print the following and fail the assert.

```
{'text': ' you'}
{'text': ' you', 'id': [1]}
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
cc @Rocketknight1 for pipelines and @ylacombe for ASR

**Comment by ylacombe** [comment]
Hey @VictorAtIfInsurance, thanks for this PR!
Could you please provide some motivation on the PR and code snippets, if necessary? 

From a quick look, I think we'd rather raise an Error when some parameters are unused, rather than just passing them through. Especially because we want to avoid silent errors or mis-usages. For example, let's say I misspelled the much-used `return_timestamps` to `retun_timestamps`. With your PR, there would be no warnings and no errors and the resulting dictionary would not be as expected!

**Comment by VictorAtIfInsurance** [comment]
Hi,

I should've marked this as WIP. The test is not setup for properly testing the intention of the PR,  it should be unused inputs and not any unused parameter as is written in the test now.

One of the reasons for this PR is inconsistent behaviour when passing chunking_length_s as a parameter to the pipeline.

E.g. I wouldv'e expected the outputs from these two different pipe calls to have the same structure of the output, i.e. that the assert would pass, 
```
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
import numpy as np


device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-large-v3-turbo"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)

data = {"raw": np.zeros(100), "sampling_rate": 16_000, "id": 1}

chunk_output = pipe(data.copy(), chunk_length_s=30)
non_chunk_output = pipe(data.copy())

print(chunk_output)
print(non_chunk_output)
assert chunk_output.keys() == non_chunk_output.keys()
```

But printing the outputs gives

```
{'text': ' you'}
{'text': ' you', 'id': [1]}
```

I.e. the unused parameter in the input "survives" when not using chunking but is removed when using chunking. 

The behaviour from the non chunking pipeline call comes from this line, so would expect something similar

https://github.com/huggingface/transformers/blob/4a173b88b50598e1dfff1759b67156f3ef257202/src/transformers/pipelines/automatic_speech_recognition.py#L472

Where the behaviour of the non chunking pipeline is especially useful if you generate your input data through a generator where you can directly call the pipeline with the generator while also retrieve metadata returned from the generator, without the extra step of first retrieving the next item, and call the pipe separately. See example:

```
def generate_data():
    iter = 0
    while iter < 10:
        yield {"raw": np.zeros(100), "sampling_rate": 16_000, "id": iter}
        iter += 1

print("Non chunking")
for output in pipe(generate_data()):
    print(output)

print()
print("Chunking")
for output in pipe(generate_data(), chunk_length_s=30):
    print(output)
```

which prints

```
Non chunking
{'text': ' you', 'id': [0]}
{'text': ' you', 'id': [1]}
{'text': ' you', 'id': [2]}
{'text': ' you', 'id': [3]}
{'text': ' you', 'id': [4]}
{'text': ' you', 'id': [5]}
{'text': ' you', 'id': [6]}
{'text': ' you', 'id': [7]}
{'text': ' you', 'id': [8]}
{'text': ' you', 'id': [9]}

Chunking
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
{'text': ' you'}
```

In these instances it wouldv'e been super useful to pass along the unused inputs.


I hope this makes sense and that you agree with my intended behaviour. I will update the tests accordingly in that case, and sorry for posting unfinished code for review.

**Comment by VictorAtIfInsurance** [comment]
@ylacombe

I've now added some code examples explaining my issue, and updated the test accordingly:)

**Review by ylacombe — COMMENTED** [review]
Hey @VictorAtIfInsurance, this actually makes sense, especially since we pass extra parameters when we don't chunk.

I'm pretty sure it doesn't align with the pipeline philosophy though, and we could maybe ask @Rocketknight1's opinion on this!

**Comment by Rocketknight1** [comment]
I think keeping the output consistent with different input kwargs definitely fits with our philosophy!

@VictorAtIfInsurance we're actually working on a project to standardize the inputs and outputs for our pipeline classes, to match an inference spec that will be common across lots of projects like TGI / transformers.js / etc. See [this PR](https://github.com/huggingface/transformers/pull/33769).

If this PR makes the pipeline always compliant with the Hub output spec, then I'm definitely in favour of it! Can you:

1) Rebase on the latest version of `main` to make sure your PR is doing the spec compliance tests
2) Add a compliance test when `chunk_length_s` is passed as well?

Other than that this PR looks good to me.

**Comment by VictorAtIfInsurance** [comment]
Hi,

Thanks for the input. 

I rebased and added check against `AutomaticSpeechRecognitionOutput` using the `compare_pipeline_output_to_hub_spec` test utility method. 

As i change the output from the pipeline call the test fails locally, which is to be expected since it doesn't follow the output structure. So this is probably not the way to go forward. Though then the non-chunk version of ASR pipeline has a problem on main, see the second assert in `test_pipeline_output_on_unused_kwargs` in #34087. 

What would be the correct way to handle invalid kwargs using the new standardized inputs for pipelines. Raise an exception such as in `test_fail_on_invalid_input_kwargs` in the same PR? #34087 

Or maybe a warning would be better in that case as it could break backwards compatibility for the users?

What would be the best next step here @ylacombe?

File: .gitignore [hunk 1]
```diff
@@ -125,6 +125,9 @@ dmypy.json
 .vs
 .vscode
 
+# devcontainer in vscode
+.devcontainer/
+
```
Comment (ylacombe) [inline .gitignore:130]: let's remove  this please

**Review by ylacombe — COMMENTED** [review]
Hey @VictorAtIfInsurance, indeed it is not compatible with `AutomaticSpeechRecognitionOutput`, that you can find [here](https://github.com/huggingface/huggingface_hub/blob/20d84915c35a541462dd416a25062bc2b2ad683d/src/huggingface_hub/inference/_generated/types/automatic_speech_recognition.py#L99-L116).

> I think keeping the output consistent with different input kwargs definitely fits with our philosophy!

@Rocketknight1, I'm not sure to follow, `AutomaticSpeechRecognitionOutput` are really constraining and some models (say `Whisper`) might return many more output fields! How can we reconcile both?

**Comment by Rocketknight1** [comment]
@ylacombe It's okay for models to return many output fields, but we're trying to get the pipelines to have standardized inputs/outputs, so that they can fit into a standard inference API.

However, in this case, the pipeline obviously has compliance issues already, as mentioned in #34087. I didn't realize this, I'm sorry! I don't think we should delay this PR because of issues in `main`, so for now, I think we can skip the compliance checks I suggested, and I'll work on making the outputs compliant later in a separate PR.

**Comment by VictorAtIfInsurance** [comment]
@ylacombe I've removed the compliance check in the test, making the test pass. Let me know if I should update anytime more here.

**Comment by VictorAtIfInsurance** [comment]
Rebased main as some tests failed due to missing files.

**Comment by VictorAtIfInsurance** [comment]
@Rocketknight1  Could we rerun the workflow? The last time it failed as I was missing some files from main, but now it should be rebased.

**Comment by VictorAtIfInsurance** [comment]
@ylacombe are we ready for merge?:)

**Review by ylacombe — APPROVED** [review]
LGTM, let's rebase on main ?

**Comment by VictorAtIfInsurance** [comment]
Sure, though there were no merge conflicts

**Review by ArthurZucker — APPROVED** [review]
Very nice thanks!

**Comment by VictorAtIfInsurance** [comment]
@ylacombe would you mind merging this PR, I do not have write access to do it myself.

**Comment by ArthurZucker** [comment]
Merged sorry for the wait!

**Pull request merged**